pkg/config/client/transport.go

```diff
@@ -0,0 +1,40 @@
+package client
+
+import (
+	"net/http"
+
+	"k8s.io/client-go/rest"
+)
+
+// AnonymousClientConfigWithWrapTransport returns a copy of the given config with all user credentials (cert/key, bearer token, and username/password) and custom transports (Transport) removed.
+// This function preserves WrapTransport for clients that care about custom HTTP behavior.
+func AnonymousClientConfigWithWrapTransport(config *rest.Config) *rest.Config {
+	newConfig := rest.AnonymousClientConfig(config)
+	newConfig.WrapTransport = config.WrapTransport
+	return newConfig
+}
+
+// NewPreferredHostRoundTripper a simple middleware for changing the destination host for each request to the provided one.
+// If the preferred host doesn't exists (an empty string) then this RT has no effect.
+func NewPreferredHostRoundTripper(preferredHostFn func() string) func(http.RoundTripper) http.RoundTripper {
+	return func(rt http.RoundTripper) http.RoundTripper {
+		return &preferredHostRT{baseRT: rt, preferredHostFn: preferredHostFn}
+	}
+}
+
+type preferredHostRT struct {
+	baseRT          http.RoundTripper
+	preferredHostFn func() string
+}
+
+func (t *preferredHostRT) RoundTrip(r *http.Request) (*http.Response, error) {
+	preferredHost := t.preferredHostFn()
+
+	if len(preferredHost) == 0 {
+		return t.baseRT.RoundTrip(r)
+	}
+
+	r.Host = preferredHost
+	r.URL.Host = preferredHost
```

Comment:
don't we have to keep the tls server name in order to verify the serving cert successfully? Or do we have localhost certs?

Reply:
yes, we do have CA that knows how to verify the serving cert for localhost at least for KCM.

Reply:
That way you could also use a node IP if necessary.

Reply:
we decided to set the `ServerName` at https://github.com/openshift/kubernetes/pull/603/commits/e5a8075fb932d3f91f8134d941d280a21149ded1#diff-d073b3d85bfc4877d354264d569dd3e19d544d1775a67f1650d661299869aa7aR448